### PR TITLE
[SPARK-5886][ML] Add StringIndexer as a feature transformer

### PR DESCRIPTION
**Pull request by mengxr**
This PR adds string indexer, which takes a column of string labels and outputs a double column with labels indexed by their frequency.

TODOs:
- [x] store feature to index map in output metadata

### CODE REVIEW

File: mllib/src/main/scala/org/apache/spark/ml/feature/LabelIndexer.scala [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+/**
+ * :: AlphaComponent ::
+ * Model fitted by [[LabelIndexer]].
+ */
+@AlphaComponent
+class LabelIndexerModel private[ml] (
+    override val parent: LabelIndexer,
+    override val fittingParamMap: ParamMap,
+    labels: Array[String]) extends Model[LabelIndexerModel] with LabelIndexerBase {
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/feature/LabelIndexer.scala]: Open-ended question: if `labels` changes from run to run (e.g. more labels are added) then the numbering may change. The label for 0 may be different from run to run. Is that going to be a surprise later? If the indexing is only transient and is never persisted anywhere, it doesn't matter, but, might it be? I'm afraid of reloading a model referring to label 0, 1, 2 that have entirely different meaning.

Maybe there are reasons this is not an issue, or the assumption is that the caller only appends values to the end of the `labels` array, ever.

Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/feature/LabelIndexer.scala]: The `labels` in a fitted model should be viewed as immutable. If there are new labels, we can train a new `LabelIndexerModel`. We are gonna put label names into the metadata, so we still tracks which index maps to which label.

We can add an option to `LabelIndexer` that allows users to specify part of the mapping. If we implement this option, then we can keep the original ordering unchanged on new datasets.

**Comment by jkbradley** [comment]
Is LabelIndexer going to be different from FeatureIndexer.  We will need a transformer to index features as well.  I've been planning to revive my old PR for DatasetIndexer which was headed this way: [https://github.com/apache/spark/pull/3000], but it would be good not to duplicate efforts.

If you want to merge the 2, that would be great, or I could push an update.  The main thing I liked about DatasetIndexer is that it could be used to choose which features to treat as continuous vs. categorical based on a maxCategories threshold.  Choosing automatically helps users to avoid having to hand-pick columns as categorical vs. continuous.

**Comment by mengxr** [comment]
The `label` in the name is quite general. It could be labels used in classification or just an arbitrary column with string labels. This is the same as `LabelEncoder` in sklearn. The `LabelIndexer` should also create ML attributes using the labels.

#3000 takes an `RDD[Vector]` and tries to decide which columns should be categorical. I don't see an overlap in terms of functionality. If the plan is to make `DatasetIndexer` a feature transformer that handles both `string labels -> indices` and `numeric -> categorical`, we can definitely go with this direction. Btw, having a single `maxCategories` might cause unexpected problems, e.g., treating categorical features as continuous. We may still need to ask users to specify which columns are categorical.

**Comment by jkbradley** [comment]
> LabelIndexer

Sounds OK, though I wonder if some people will think it only applies to "labels" since we use that term for prediction.

> String vs. numeric

Oops, I didn't realize this was taking String.  (I had only looked at the JIRA and PR descriptions, not the code.)  They sound different.

**Comment by mengxr** [comment]
This PR contains changes in #5171.

**Comment by mengxr** [comment]
test this please

**Comment by mengxr** [comment]
@srowen The output column now contains ML attributes. Do you have time to make another pass?

File: mllib/src/main/scala/org/apache/spark/ml/feature/LabelIndexer.scala [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+  /** @group setParam */
+  def setOutputCol(value: String): this.type = set(outputCol, value)
+
+  // TODO: handle unseen labels
+
+  override def fit(dataset: DataFrame, paramMap: ParamMap): LabelIndexerModel = {
+    val map = this.paramMap ++ paramMap
+    val counts = dataset.select(map(labelCol)).map(_.getString(0)).countByValue()
+    val labels = counts.toSeq.sortBy(-_._2).map(_._1).toArray
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/feature/LabelIndexer.scala]: Instead of ordering by count, which requires all that counting, can labels just be ordered alphabetically? maybe I miss why it's important to sort them this way, but I'd think it's arbitrary. A deterministic function of the labels themselves might be less surprising or something later, though I can't name a specific problem it would solve, now.

File: mllib/src/main/scala/org/apache/spark/ml/feature/LabelIndexer.scala [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+    val indexer = udf { label: String =>
+      if (labelToIndex.contains(label)) {
+        labelToIndex(label)
+      } else {
+        throw new SparkException(s"Unseen label: $label.")
+      }
+    }
+    val outputColName = map(outputCol)
+    val metadata = NominalAttribute.defaultAttr
+      .withName(outputColName).withValues(labels).toStructField().metadata
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/feature/LabelIndexer.scala]: Yep so here is where the labels go into the metadata, in order. That's good.

File: mllib/src/main/scala/org/apache/spark/ml/feature/LabelIndexer.scala (hunk 2 above)
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/feature/LabelIndexer.scala]: Since we need to distinct labels anyway, there is little overhead with count and sort. Couple benefits:

1. better sparsity as there will be more zeros
2. easy inspection as common labels are at the beginning.

Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/feature/LabelIndexer.scala]: Hm because the most common label is encoded as 0.0? it feels like a funny optimization since it's not "really" a 0, and these values are handled as individual values or dense vectors mostly (right?). Is counting the whole data set by label really that cheap? it still means looking at every datum even if there are few labels. I don't strongly object or anything, this just hadn't occurred to me.

Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/feature/LabelIndexer.scala]: In vector assembler, where we merge multiple columns into a vector column, the encoded indices are stored inside a sparse vector. For multiclass labels, this doesn't help much. But for binary labels, this saves at least half of the storage.

`countByValue` costs about the same as `distinct`:
https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/rdd/RDD.scala#L1041
https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/rdd/RDD.scala#L313

Comment (petro-rudenko) [inline mllib/src/main/scala/org/apache/spark/ml/feature/LabelIndexer.scala]: Maybe it makes sense to use implementation from [DatasetIndexer](https://github.com/apache/spark/pull/3000/) or vise versa, so to keep the transformation logic in one place, and if there would be need to optimize performance (e.g. get the advantages of columnar storages - some storages can provide column cardinality metadata) - to change it in one place.

File: mllib/src/main/scala/org/apache/spark/ml/feature/LabelIndexer.scala [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+ * :: AlphaComponent ::
+ * A label indexer that maps a string column of labels to an ML column of label indices.
+ * The indices are in [0, numLabels), ordered by label frequencies.
+ * So the most frequent label gets index 0.
+ */
+@AlphaComponent
+class LabelIndexer extends Estimator[LabelIndexerModel] with LabelIndexerBase {
+
+  /** @group setParam */
+  def setLabelCol(value: String): this.type = set(labelCol, value)
```
Comment (petro-rudenko) [inline mllib/src/main/scala/org/apache/spark/ml/feature/LabelIndexer.scala]: If it suppose to be general indexer, not just for label column, maybe it makes sense to call it ColumnIndexer and use setInputCol instead.

Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/feature/LabelIndexer.scala]: I agree with this since we have used "label" to mean label for prediction thus far.  (But @mengxr disagreed based on precedence from sklearn; you can see the comments above.)

I also added questions in the JIRA about whether this class should support multiple input and output columns and support more native types than String.

Comment (petro-rudenko) [inline mllib/src/main/scala/org/apache/spark/ml/feature/LabelIndexer.scala]: Yes, but in jira's example it uses setInputCol rather then setFeatureCol:
```scala
val i = new LabelIndexer()
  .setInputCol("country")
  .setOutputCol("countryIndex")
```
which makes more sense as for me.

Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/feature/LabelIndexer.scala]: I agree

Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/feature/LabelIndexer.scala]: I will rename `labelCol` to `inputCol`. Feel free to suggest better class names. I'm okay with not calling it `LabelIndexer`. How about `StringToNominal`?

Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/feature/LabelIndexer.scala]: Any thoughts about generalizing it beyond String?

File: mllib/src/main/scala/org/apache/spark/ml/feature/LabelIndexer.scala [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+
+/**
+ * Base trait for [[LabelIndexer]] and [[LabelIndexerModel]].
+ */
+private[feature] trait LabelIndexerBase extends Params with HasLabelCol with HasOutputCol {
+
+  /** Validates and transforms the input schema. */
+  protected def validateAndTransformSchema(schema: StructType, paramMap: ParamMap): StructType = {
+    val map = this.paramMap ++ paramMap
+    val labelType = schema(map(labelCol)).dataType
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/feature/LabelIndexer.scala]: This will be shorter if you use the utility methods checkInputColumn and addOutputColumn.

File: mllib/src/main/scala/org/apache/spark/ml/feature/LabelIndexer.scala (hunk 4 above)
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/feature/LabelIndexer.scala]: For model persistence, we need to save labels to disk. String would be ideal and may be sufficient.

File: mllib/src/main/scala/org/apache/spark/ml/feature/LabelIndexer.scala [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+  /** @group setParam */
+  def setOutputCol(value: String): this.type = set(outputCol, value)
+
+  override def transform(dataset: DataFrame, paramMap: ParamMap): DataFrame = {
+    val map = this.paramMap ++ paramMap
+    val indexer = udf { label: String =>
+      if (labelToIndex.contains(label)) {
+        labelToIndex(label)
+      } else {
+        throw new SparkException(s"Unseen label: $label.")
```
Comment (petro-rudenko) [inline mllib/src/main/scala/org/apache/spark/ml/feature/LabelIndexer.scala]: Maybe make an option to specify default value for unknown label rather then throwing exception.

File: mllib/src/main/scala/org/apache/spark/ml/feature/LabelIndexer.scala (hunk 4 above)
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/feature/LabelIndexer.scala]: Since the data are already in a DataFrame, we know we can save it to disk.  But keeping it simple with String is fine for now.

Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/feature/LabelIndexer.scala]: I rename the class to `StringIndexer` and changed `labelCol` to `inputCol`.

File: mllib/src/main/scala/org/apache/spark/ml/feature/LabelIndexer.scala (hunk 6 above)
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/feature/LabelIndexer.scala]: I left a TODO for this. The issues are

1. if we assign unseen label a unique index, it creases the number of values by 1 and a binary feature is no longer binary.
2. if we map unseen label to an exiting index, it may silently change the behavior and users may not be aware of it.
3. another option is to map the unseen label to `Double.NaN`, which is treated as missing value and can be filled in later.

So I think we can discuss this in a follow-up PR.

**Comment by mengxr** [comment]
Merged into master.
